Client/lib/gMap.js

```diff
@@ -0,0 +1,282 @@
+        destination : gMap.pathLatLng[gMap.pathLatLng.length-1], // can be latLng or string (this is required)
+        travelMode : google.maps.TravelMode[gMap.travelMode],
+        waypoints : waypoints
+    };
+
+    gMap.directionsService.route(request, function(response, status) {
+      if (status === google.maps.DirectionsStatus.OK) {
+        gMap.directionsDisplay.setDirections(response);
+        gMap.emptyMarkers();
+        gMap.makeMarker(response.routes[0].legs[0].start_location);
+        for (var i = 0; i < response.routes[0].legs.length; i++) {
+          gMap.makeMarker(response.routes[0].legs[i].end_location);
+          gMap.distance[i] = response.routes[0].legs[i].distance.value;
+          gMap.duration[i] = response.routes[0].legs[i].duration.value;
+        };
+        oldPath.setMap(null);
+      }else{
+        gMap.pathLatLng.pop(); //remove the last waypoint added
+      }
+      google.maps.event.addListener(gMap.directionsDisplay, 'directions_changed', function() {
+        if(gMap.directionsDisplay.directions.routes){
+          gMap.pathLatLng[0] = gMap.directionsDisplay.directions.routes[0].legs[0].start_location
+        }
+        for (var i = 1; i <= gMap.directionsDisplay.directions.routes[0].legs.length; i++) {
+          gMap.pathLatLng[i] = gMap.directionsDisplay.directions.routes[0].legs[i-1].end_location
+        };
+        gMap.emptyMarkers();
+        gMap.createPath();
+      });
+    });
+  }
+
+  gMap.makeMarker = function (latLng){
+    var markerImage = {
+      url: 'https://cdn2.iconfinder.com/data/icons/snipicons/500/map-marker-512.png',
+      scaledSize: new google.maps.Size(40,40),
+      origin: new google.maps.Point(0,0),
+      anchor: new google.maps.Point(20,50)
+    };
+
+    var index = gMap.markers.length;
+    var marker = new google.maps.Marker({
+      map: gMap.map,
+      icon: markerImage,
+      position: latLng,
+      title: 'marker ' + index
+    });
+    gMap.markers.push(marker);
+    google.maps.event.addListener(marker, 'click', function() {
+      console.log(index);
+    });
+  }
+  gMap.emptyMarkers = function (){
+    for (var i = 0; i < gMap.markers.length; i++) {
+      gMap.markers[i].setMap(null);
+    };
+    gMap.markers.length = 0;
+  }
+
+  gMap.getGeolocation = function (callback){
+    var pos = new google.maps.LatLng(-33.73, 149.02);
+    if (navigator.geolocation) {
+      navigator.geolocation.getCurrentPosition(function(position){
+        //console.log(position)
+        pos = new google.maps.LatLng(position.coords.latitude, position.coords.longitude);
+        callback(pos);
+      });
+    }else{
+      callback(pos);
+    }
+  }
+  gMap.select = function (index){
+    var obj = {};
+    obj.title = gMap.markers[index].title;
+    obj.position = {};
+    obj.position.lat = gMap.markers[index].getPosition().lat();
+    obj.position.lng = gMap.markers[index].getPosition().lng();
+    return obj;
+  }
+  gMap.remove = function (index){
```

Comment:
This function is a bit confusing as it does not look like what I expect a remove to be. In addition to performing a pop() you seem to be splicing out an element from the middle and then pushing it onto the end.